packages/remirror__extension-entity-reference/src/entity-reference-extension.ts

```diff
@@ -0,0 +1,264 @@
+    const idAttr = `data-entityReference`;
+
+    return {
+      ...override,
+      excludes: '',
+      attrs: {
+        ...extra.defaults(),
+        id: { default: '' },
+      },
+      // Note that we don't implement `ParseDom`.
+      // The reasoning behind this is to disable pasting (parsing) the entityReference marks in the editor.
+      toDOM: (mark: Mark) => [
+        'span',
+        {
+          [idAttr]: mark.attrs.id,
+        },
+        0,
+      ],
+    };
+  }
+
+  /**
+   * Create the extension plugin for inserting decorations into the editor.
+   */
+  createPlugin(): CreateExtensionPlugin {
+    return {
+      state: {
+        init: (_: { [key: string]: any }, state: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(state.doc);
+          return { entityReferences };
+        },
+        apply: (_tr: Transaction, _value: any, _oldState: EditorState, newState: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(newState.doc);
+          return { entityReferences };
+        },
+      },
+      props: {
+        decorations: (state: EditorState) => {
+          return createDecorationSet({ state, extension: this });
+        },
+      },
+    };
+  }
+
+  @command()
+  addEntityReference(id?: string): CommandFunction {
+    return ({ state, tr, dispatch }) => {
+      const { from, to } = state.selection;
+      const newId = id ?? (this.options.createId ? this.options.createId() : uniqueId());
+      const newMark = this.type.create({
+        id: newId,
+      });
+      try {
+        tr = tr.addMark(from, to, newMark);
+      } catch (error: any) {
+        throw new Error(
+          `Can't add entityReference ${JSON.stringify(
+            newMark,
+          )} to range {from: ${from}, to: ${to}}: ${error.message}`,
+        );
+      }
+
+      dispatch?.(tr);
+      return true;
+    };
+  }
+
+  @command()
+  removeEntityReference(entityReferenceId: string): CommandFunction {
+    return ({ tr, dispatch }) => {
+      if (!entityReferenceId) {
+        throw new Error(`EntityReference can't be removed without specifying tags its ID!`);
+      }
+
+      const singleMark = this.type.create({ id: entityReferenceId });
+      try {
+        // Remove all entityReference marks with id=entityReferenceId in all the document
+        // Getting the last position https://discuss.prosemirror.net/t/getting-the-last-node-and-its-pos-in-the-document/3091
+        tr = tr.removeMark(1, tr.doc.content.size, singleMark);
+      } catch (error: any) {
+        throw new Error(`Can't remove entityReference ${entityReferenceId}: ${error.message}`);
+      }
+      dispatch?.(tr);
+      return true;
+    };
+  }
+
+  @command()
+  redrawEntityReferences(): CommandFunction {
+    return ({ tr, dispatch }) => {
+      dispatch?.(
+        tr.setMeta(EntityReferenceExtension.name, {
+          type: ActionType.REDRAW_ENTITY_REFERENCES,
```

Comment:
The reason behind this command is to force update the editor (e.i recalculate decorations) outside of the editor. Is there a better way to do this?

Reply:
Ah apologies @IdrissMahjoubi, I was expecting to see something in the plugin listening to this meta type? It may be possible to use the existing [`forceUpdate`](https://github.com/remirror/remirror/blob/main/packages/remirror__core/src/builtins/commands-extension.ts#L520-L526) command to fulfil this use case?

---

P.s. are decorations necessary anymore? As marks already get rendered into the document?

I wonder if `addEntityReference` should also allow for a `color` mark attribute? To determine how it is rendered in document?

P.s. If marks can be used in this way, it might be worth looking at the CSS property [`mix-blend-mode`](https://developer.mozilla.org/en-US/docs/Web/CSS/mix-blend-mode) to determine the color behaviour when there are overlapping entity references

cc @ronnyroeller for context on color overlapping behaviour.

Reply:
TODO:
- Check if `forceUpdate` can do the same
- If not: Add a comment to the command to explain what it does (force redraw if an external color for a entity is changed) and how it works (inject a change to the document) and why this is required to be implemented like that

> I wonder if addEntityReference should also allow for a color mark attribute?

This might work for simple use case but wouldn't work for what we need:
- Our users can change the color via the entity. -> This would require us to change the color in all mark attributes across all Prosemirror documents, whenever the color of an entity changes.
- Our entities have semantic colors, which are then translated to hex colors based on light/dark mode. -> This would require us to change the color in all mark attributes across all Prosemirror documents, whenever the user switches between light/dark mode.

So, therefore I'd keep `color` out of scope for now (to keep the extension as simple as possible). Maybe somebody could add in the future, if there is a real use for it.